helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
+                .current_dir(grammar_dir)
+                .output()
+                .expect("Failed to execute 'git'");
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+        }
+    };
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) {
+    if !Command::new("git")
+        .args(["remote", "set-url", REMOTE_NAME, remote_url])
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'")
+        .status
+        .success()
+        && !Command::new("git")
+            .args(["remote", "add", REMOTE_NAME, remote_url])
+            .current_dir(repository)
+            .output()
+            .expect("Failed to execute 'git'")
+            .status
+            .success()
+    {
+        eprintln!("Failed to set remote '{}'", remote_url);
+    }
+}
+
+fn get_repository_info(repository: &Path, args: Vec<&str>) -> Option<String> {
+    let output = Command::new("git")
+        .args(args)
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'");
+    if output.status.success() {
+        let mut remote = String::from_utf8_lossy(output.stdout.as_slice()).into_owned();
+        // remove trailing newline
+        remote.pop();
+        Some(remote)
+    } else {
+        None
+    }
+}
+
+fn build_grammar(grammar: GrammarConfiguration) {
+    let grammar_dir = if let GrammarSource::Local { ref path } = grammar.source {
+        PathBuf::from(path)
+    } else {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("../runtime/grammars/sources")
+            .join(grammar.grammar_id.clone())
+    };
+
+    if grammar_dir.read_dir().is_err() {
+        eprintln!(
+            "The directory {:?} is empty, you probably need to use 'hx --fetch-grammars'?",
+            grammar_dir
+        );
+        std::process::exit(1);
+    }
+
+    let path = match grammar.path {
+        Some(ref subpath) => grammar_dir.join(subpath),
+        None => grammar_dir,
+    }
+    .join("src");
+
+    build_library(&path, grammar).unwrap();
+}
+
+// Returns the user-defined set of grammars if the user has a languages.toml,
+// defaulting to the built-in languages.toml entries.
```

Comment:
I understand the purpose of this is to enable opting out of a language or many, but if I'm understanding this correctly, is a consequence of this that if the user has any language config at all that they will lose all the default grammars unless they copy them all into their user config? If so, this seems like a poor user experience.

Reply:
Oh whoops, I need to revert my changes to `helix_core::config`. With the addition of `use-grammars`, now we use the fully merged config as usual